### ISSUE TITLE
[Feature Request] Use Embedding for Discrete features

### ISSUE THREAD

**Issue by zippeurfou** [DLR-RM/stable-baselines3#1210]
### 🚀 Feature

Use Embedding and Embedding bags if the Dict space contains Discrete or MultiDiscrete for the MultiInput policy or MLPpolicy.

### Motivation

I was wondering how stable baseline treats different space types as inputs in the case of multiInput policies. I was happy to see that any input that is interpreted as an image go into a CNN (https://github.com/DLR-RM/stable-baselines3/blob/e39bc3da00c49413b765176af1b95f2361a35098/stable_baselines3/common/torch_layers.py#L261).
It would be great to apply this same kind of magic if we observe different space type.
More specifically I think Discrete and MultiDiscrete could be pretty straight forward ones. We could use [Google rule of thumb](https://developers.googleblog.com/2017/11/introducing-tensorflow-feature-columns.html) -> embedding_dimensions = number_of_categories**0.25 for the embedding size and throw this to the network.
For multidiscrete we can just use embedding bags

### Pitch

Edit  `CombinedExtractor` to detect if the space is Discrete or MultiDiscrete.
If so, apply a preprocessing that add respectively an Embedding and an EmbeddingBag to the input of size  number_of_categories**0.25

### Alternatives

Alternative could be to allow the user via the policy parameter to specify the rule to take for Discrete, for example providing a module list to input.

### Additional context

NA

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo

**Comment by qgallouedec** [comment]
Thank you for this suggestion. Why would this only apply to MultiInput containing Dicrete subspace and not also to simple Discrete inputs?

**Comment by zippeurfou** [comment]
Thanks @qgallouedec. I think I wasn’t clear. I meant embedding for discrete and embedding bags for multi discrete

**Comment by qgallouedec** [comment]
I want to make sure I understand your suggestion. 
For example, in the case of PPO/CartPole, is the default mode you propose equivalent to what the following code does in the current version of SB3?

```python
import gym
import numpy as np

from stable_baselines3 import PPO

env = gym.make("CartPole-v1")

embedding_size = int(np.ceil(env.action_space.n**0.25))

PPO("MlpPolicy", env, policy_kwargs=dict(net_arch=[64, embedding_size]))
```

**Comment by araffin** [comment]
Hello,
I think @zippeurfou is referring to embeddings à la Word2Vec (so a one-hot encoding is transformed into a feature vector): https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html ?

@zippeurfou are you aware we are already pre-processing input to one-hot vector automatically?
I'm also not sure if it would make sense as compared to NLP, the "vocabulary size" (here the number of possible states) is usually not that high in RL (for instance, for Atari game, all the buttons combination). 

Do you have any reference of RL paper that apply this trick?
(if not, then I would prefer to have some quantitative results before making such strategy the default)

> Alternatives

a custom feature extractor/policy with an Embedding layer would be a good alternative?

**Comment by zippeurfou** [comment]
> @zippeurfou are you aware we are already pre-processing input to one-hot vector automatically?
No I didn't go as far in the source code to it. But now I am digging into the source code it indeed does as you said (https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/preprocessing.py#L106-L118).
But since you do that and then add a linear to it (https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/torch_layers.py#L135) + activation. You're basically doing an embedding (embedding doesn't have the activation thought). You also don't need an embedding bag there since you can let the network figure it out. 
The only advantage of having an embedding would be compute speed but then this might not be too much of an issue still. 
I think we can close this one.
Thanks to both of you.
